### PR TITLE
Improve kf set values

### PR DESCRIPTION
**Pull request by subodh101**
What?
- Remove make rule `set-values`.
- Add zone variable in Kptfile
- Add value validation step

Why?
- Deploy kubeflow with five necessary variables MGMTCTXT, KUBEFLOW-NAME, LOCATION, ZONE, PROJECT
- Just like management cluster. Ask the user to set values by `kpt` command in docs

Ref: https://github.com/kubeflow/gcp-blueprints/pull/135#issuecomment-692113900

### CODE REVIEW

**Comment by subodh101** [comment]
Test will fail, Wating on https://github.com/kubeflow/gcp-blueprints/pull/135

**Comment by jlewi** [comment]
Nice change!
/lgtm
/hold 

@Bobgy Any feedback?

File: kubeflow/instance/Kptfile [hunk 1]
```diff
@@ -37,6 +37,11 @@ openAPI:
         setter:
           name: location
           value: LOCATION
+    io.k8s.cli.setters.zone:
+      x-k8s-cli:
+        setter:
+          name: zone
```
Comment (Bobgy) [inline kubeflow/instance/Kptfile:43]: There seems to be a mismatch, zone parameter should be set by
```
kpt cfg set ./upstream/manifests/gcp gcloud.compute.zone $(ZONE)
```

So maybe it should look like:
```
io.k8s.cli.setters.gcloud.compute.zone:
```
?

I also have a concern that we will eventually default to regional clusters, can we change the parameter to location, but document currently we only support zone here.

**Review by Bobgy — COMMENTED** [review]
/cc @jlewi

File: kubeflow/instance/Kptfile (hunk 1 above)
Comment (jlewi) [reply kubeflow/instance/Kptfile:43]: I think we should potentially have setters for both location and gcloud.compute.zone. 

Which setter gets used in a particular location should then depend on whether the value must be zonal  (use gcloud.compute.zone) or can be regional or zonal (use location).

Most of the current usages can use either (location) e.g. cluster location.

I think one place where we have a zonal resource is the PD backing the KFP MySQL.

File: kubeflow/Makefile [hunk 2]
```diff
@@ -28,28 +30,13 @@ BUILD_DIR=.build
...
-	kpt cfg set ./upstream/manifests/stacks/gcp gcloud.core.project <PROJECT_TO_DEPLOY_IN>
-
-	kpt cfg set ./instance name <YOUR_KF_NAME>
-	kpt cfg set ./instance location <YOUR_REGION or ZONE>
-	kpt cfg set ./instance gcloud.core.project <YOUR PROJECT>
-	kpt cfg set ./instance email <YOUR_EMAIL_ADDRESS>
+# Validate cluster values are changed from default dummy values
+.PHONY: validate-values
+validate-values:
+	@if [[ $(MGMTCTXT) == "MANAGEMENT-CTXT" || $(NAME) == "KUBEFLOW-NAME" || $(LOCATION) == "LOCATION" || $(ZONE) == "ZONE" || $(PROJECT) == "PROJECT" ]]; then\
```
Comment (Bobgy) [inline kubeflow/Makefile:36]: You have LOCATION here, but you didn't init its value

Comment (subodh101) [reply kubeflow/Makefile:37]: Yes, it was present in the previous PR. Now available.

File: kubeflow/instance/Kptfile (hunk 1 above)
Comment (Bobgy) [reply kubeflow/instance/Kptfile]: When we move to a regional compatible setup, IMO there shouldn't be any zonal resources left. (e.g. we should replace those PDs with gcs and regional/zonal cloudsql)
But we can leave this to a later topic.

**Comment by Bobgy** [comment]
/lgtm
/approve
Thank you for the great contribution!

**Comment by jlewi** [comment]
@subodh101 apologies that this didn't get merged. Looks like there was a hold that wasn't removed.

/hold cancel

**Comment by subodh101** [comment]
/retest

Last [test](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubeflow_gcp-blueprints/137/kubeflow-gcp-blueprints-presubmit/1308300383084351488) ran just fine.

**Comment by jlewi** [comment]
@subodh101 Looks like the tests keep failing; have you looked at the logs to see what the error is?

**Comment by subodh101** [comment]
Yes. I checked the logs.
Unfortunately, I couldn't see anything in the logs. `auto-deploy/deploy-kf-master` is failing, I can't see the container logs. Are those logs available somewhere public?
I am starting to realize that this problem started occurring after https://github.com/kubeflow/gcp-blueprints/pull/134 went in. Could it be a test is written to check `../hack/create_context.sh`?

**Comment by subodh101** [comment]
@jlewi Could you drop a "/test all" to check my suspicion?

**Comment by subodh101** [comment]
Everything in this PR seems okay to me. There should definitely a problem with upstream.

Could you explain the purpose of deleting something that doesn't seem to exist?

https://github.com/kubeflow/manifests/blob/c728bc737cfadc2816ecd50c5c96ee59cd1d9b1a/tests/Makefile#L41-L46

After running above `make gcp-reset` (also maybe `yq` command is incorrect). It generates a diff (pretty sure that's not intended) in the `gcp/Kptfile`.

```zsh
           values:
-          - marker: ${name}
-            ref: '#/definitions/io.k8s.cli.setters.name'
+            - marker: ${name}
+              ref: '#/definitions/io.k8s.cli.setters.name'```
```

Ref.: https://googlecontainertools.github.io/kpt/guides/producer/substitutions/#example

**Comment by jlewi** [comment]
@subodh101 You can read about how to access test logs here:
https://github.com/kubeflow/testing#working-with-the-test-infrastructure

**Comment by jlewi** [comment]
Is the diff in the KptFile just a change in indentation?

The motivation behind deleting the "isSet" value is because I think if isSet is true it might prevent `kpt` from doing automatic substitution of certain values like project. That was my theory at least haven't verified.

Its possible that ends up deleting something that doesn't exist and its an error.

**Comment by jlewi** [comment]
It looks like the prow job is exiting with an error:

```
ERROR|2020-10-10T11:04:42|/src/kubeflow/testing/py/kubeflow/testing/tekton_client.py|46| KeyError: 'status'
Traceback (most recent call last):
  File "/src/kubeflow/testing/py/kubeflow/testing/tekton_client.py", line 41, in log_status
    workflow["status"]["conditions"][0]["reason"])
KeyError: 'status'
INFO|2020-10-10T11:04:42|/src/kubeflow/testing/py/kubeflow/testing/tekton_client.py|88| Retry on exception: 'status'; stack trace:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/retrying.py", line 200, in call
    attempt = Attempt(fn(*args, **kwargs), attempt_number, False)
  File "/src/kubeflow/testing/py/kubeflow/testing/tekton_client.py", line 119, in get_namespaced_custom_object_with_retries
    condition = result["status"]["conditions"][0]["reason"]
KeyError: 'status'
```

The AutoDeploy Tekton pipeline failed
https://kf-ci-v1.endpoints.kubeflow-ci.cloud.goog/tekton/#/namespaces/auto-deploy/pipelineruns/deploy-kf-master-mmbgs

The test timed out waiting for the cluster to be ready

```
kubectl --context=kf-ci-management wait --for=condition=Ready --timeout=600s containercluster kf-vbp-1010-3dd
error: timed out waiting for the condition on containerclusters/kf-vbp-1010-3dd
```

Full logs
[kubeflow-ci.value.txt](https://github.com/kubeflow/gcp-blueprints/files/5371637/kubeflow-ci.value.txt)


This looks like a quota issue kubeflow/testing#758

**Comment by subodh101** [comment]
> Is the diff in the KptFile just a change in indentation?

yes. It only makes the indentation change but at many places. 


@jlewi Thank you for debugging! :)
Now that there won't be any quota https://github.com/kubeflow/testing/issues/758#issuecomment-709280888 issue.
It should not timeout.
Please run the test once. I will try to investigate it if fails again.

**Comment by jlewi** [comment]
Looks like the tests are still failing.

**Comment by subodh101** [comment]
Could you add me as a ci viewer https://github.com/kubeflow/internal-acls/pull/356?

**Comment by subodh101** [comment]
Screen Shot 2020-10-17 at 20 26 14

Argo UI was also not working.

File: kubeflow/Makefile [hunk 3]
```diff
@@ -281,7 +267,17 @@ acm-kubeflow: hydrate-asm hydrate-kubeflow
...
+clean-build: validate-values
+	# Dedupe values to ./upstream/manifest
+	kpt cfg set ./upstream/manifests/gcp name $(NAME)
+	kpt cfg set ./upstream/manifests/gcp gcloud.core.project $(PROJECT)
+	kpt cfg set ./upstream/manifests/gcp gcloud.compute.zone $(ZONE)
+	kpt cfg set ./upstream/manifests/gcp location $(LOCATION)
+	kpt cfg set ./upstream/manifests/gcp log-firewalls $(FIREWALL_LOGS)
+
+	kpt cfg set ./upstream/manifests/stacks/gcp name $(NAME)
+	kpt cfg set ./upstream/manifests/stacks/gcp gcloud.core.project $(PROJECT)
```
Comment (subodh101) [inline kubeflow/Makefile:279]: This is creating a problem in the tests.

From logs:

```
# Dedupe values to ./upstream/manifest
2020-10-10T15:21:40.157842566Z  deploy-kf-master-mmbgs-deploy-gcp-gwrzl-pod-5fgkj  step-deploy-gcp                       /bin/sh: 1: kubeflow-ci-deployment: not found
2020-10-10T15:21:40.157838419Z  deploy-kf-master-mmbgs-deploy-gcp-gwrzl-pod-5fgkj  step-deploy-gcp                       /bin/sh: 1: ZONE: not found
2020-10-10T15:21:40.157834159Z  deploy-kf-master-mmbgs-deploy-gcp-gwrzl-pod-5fgkj  step-deploy-gcp                       /bin/sh: 1: us-central1-c: not found
2020-10-10T15:21:40.157829567Z  deploy-kf-master-mmbgs-deploy-gcp-gwrzl-pod-5fgkj  step-deploy-gcp                       /bin/sh: 1: kf-vbp-1010-3dd: not found
2020-10-10T15:21:40.157822154Z  deploy-kf-master-mmbgs-deploy-gcp-gwrzl-pod-5fgkj  step-deploy-gcp                       /bin/sh: 1: [[: not found
```

```
kpt cfg set ./upstream/manifests/gcp gcloud.compute.zone ZONE
set 3 fields
set 84 fields
```

No value of ZONE could also be a problem.

**Comment by subodh101** [comment]
Will add the ZONE back once tests https://github.com/kubeflow/testing/pull/759 are updated.

Previous validation-values makerule was running on the local machine but not inside the test environment.

```
/bin/sh: 1: [[: not found
/bin/sh: 1: kf-vbp-1010-3dd: not found
...
```

Running the validation-values commands inside shell fixes the problem.

I suspect the last test failed because of quota problem.

**Comment by jlewi** [comment]
@subodh101 Groups should now be sync'd automatically so any group proposals you submitted should have been applied.

**Comment by subodh101** [comment]
For the [group sync](https://github.com/kubeflow/internal-acls/blob/master/google_groups/ci-viewer.members.txt), I don't think I got the viewer access to projects `kubeflow-ci` and `kubeflow-ci-deployment` for the specified email id.
Maybe I need to add [myself again](https://github.com/kubeflow/internal-acls/pull/367/files#diff-00db984b348683011e137796cf4265db834a8ba3c10a0636e5ae23c5f327c38aR29-R32) after your PR of autosync is merged.

**Comment by subodh101** [comment]
Could you also add me [manually](https://github.com/kubeflow/internal-acls/tree/master/google_groups#internal-acls)? 
[ASK] I am expecting that the above two projects will pop up in my gcloud console. Please let me know if that's not the case.
cc: @jlewi

**Comment by jlewi** [comment]
Apologies as you noticed it looks like your changes got lost in the migration. I confirmed subodh.pushkar@gmail.com is now a member of ci-viewer. Please take a look and see if its working for you.

**Comment by subodh101** [comment]
Yes, I can see the logs now. Thank you.

**Comment by Bobgy** [comment]
@subodh101 I'll investigate how to fix the presubmit test, it's broken for everyone.

**Comment by jlewi** [comment]
kubeflow/gcp-blueprints#143

**Comment by jlewi** [comment]
Per #143 Tekton pipelineruns appear to be running again so lets try again.

/test all

**Comment by jlewi** [comment]
On the most recent test failure.

```
kubectl --context=kf-ci-management wait --for=condition=Ready --timeout=600s  containercluster kf-vbp-1029-4e8
error: timed out waiting for the condition on containerclusters/kf-vbp-1029-4e8
Makefile:304: recipe for target 'wait-gcp' failed
make: *** [wait-gcp] Error 1
```

The auto-deployments from master don't appear to be healthy:
https://kf-ci-v1.endpoints.kubeflow-ci.cloud.goog/auto_deploy/

So something may be broken on master.

**Comment by subodh101** [comment]
Test passed 🚀 . Thank you so much!!

**Comment by jlewi** [comment]
Its green! Woo Hoo!
@subodh101 Thanks for your patience.

**Comment by Bobgy** [comment]
/unhold
/lgtm
Thank you @subodh101 so much!

**Comment by Bobgy** [comment]
> Before we merge this I want to fix over the ZONE variable.
> However, there is an error of `No space left on device` in the [kubeflow/testing#759](https://github.com/kubeflow/testing/pull/759).
> This is present in many PRs getting merged as well.

Sorry I might have missed this statement, feel free to fix the ZONE variable in a new PR.

**Pull request merged**